### PR TITLE
OpenSearch - Hello LTR Update

### PR DESCRIPTION
**Pull request by macohen**
By way of learning LTR with OpenSearch, I updated a bunch of code here to use OpenSearch. It's not done, but I'm opening as a draft just for some early feedback.

### CODE REVIEW

**Comment by macohen** [comment]
I've been able to make hello-ltr work with OpenSearch 2.5.0 and the ltr plugin released here: https://github.com/opensearch-project/opensearch-learning-to-rank-base/releases/tag/release-v2.1.0

File: docker-compose.yml [hunk 1]
```diff
@@ -33,6 +33,28 @@ services:
       ELASTICSEARCH_URL: "http://hello-ltr-elastic:9200"
       SERVER_HOST: "0.0.0.0"
 
+  opensearch:
+    build:
+      context: ./notebooks/opensearch/.docker/opensearch-docker/
+      dockerfile: Dockerfile
+    container_name: hello-ltr-opensearch
+    ports:
+      - 9201:9201
```
Comment (mkr) [inline docker-compose.yml:42]: I think we also need to disable security here, i.e. something like
```suggestion
    environment:
      - "OPENSEARCH_JAVA_OPTS=-Xms512m -Xmx512m"
      - "DISABLE_INSTALL_DEMO_CONFIG=true"
      - "DISABLE_SECURITY_PLUGIN=true"
      - "discovery.type=single-node"
```

File: docker-compose.yml [hunk 2]
```diff
@@ -33,6 +33,28 @@ services:
...
+
```
Comment (mkr) [inline docker-compose.yml:49]: I think we also need to disable security here, i.e. something like
```suggestion
    
    environment:
      - "OPENSEARCH_JAVA_OPTS=-Xms512m -Xmx512m"
      - "DISABLE_INSTALL_DEMO_CONFIG=true"
      - "DISABLE_SECURITY_PLUGIN=true"
      - "discovery.type=single-node"
```

File: docker-compose.yml [hunk 3]
```diff
@@ -33,6 +33,28 @@ services:
...
+    container_name: hello-ltr-osd
+    expose:
+      - "5602"
+    ports:
+     - "5602:5602"
+    environment:
+      OPENSEARCH_HOSTS: "http://hello-ltr-opensearch:9201"
+      OPENSEARCH_URL: "http://hello-ltr-opensearch:9201"
+      SERVER_HOST: "0.0.0.0"
+
```
Comment (mkr) [inline docker-compose.yml:65]: ... and here
```suggestion
      DISABLE_SECURITY_DASHBOARDS_PLUGIN: true
```

File: ltr/client/opensearch_client.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+
+        In the future, we may wish to isolate an Index's feature
+        store to a feature store of the same name of the index
+    """
+    def __init__(self, configs_dir='.'):
+        self.docker = os.environ.get('LTR_DOCKER') != None
+        self.configs_dir = configs_dir #location of elastic configs
+
+        if self.docker:
+            self.host = 'opensearch-node1'
```
Comment (mkr) [inline ltr/client/opensearch_client.py:49]: This would be `opensearch` when using the root docker-compose.yml?

**Comment by mkr** [comment]
Thank you @macohen, would you mind adding OpenSearch also to the README.md in the root of the project?

File: ltr/client/opensearch_client.py [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+import requests
+
+from .base_client import BaseClient
+from ltr.helpers.handle_resp import resp_msg
+
+import json
+from opensearchpy import OpenSearch
+from opensearchpy import helpers
+
+class ElasticResp():
```
Comment (david-fisher) [inline ltr/client/opensearch_client.py]: Might this be better named OpenSearchResp?

File: ltr/client/opensearch_client.py [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+    def check_index_exists(self, index):
+        return self.es.indices.exists(index=index)
+
+    def delete_index(self, index):
+        resp = self.es.indices.delete(index=index, ignore=[400, 404])
+        resp_msg(msg="Deleted index {}".format(index),
+                 resp=ElasticResp(resp), throw=False, ignore=[400, 404])
+
+    def create_index(self, index):
+        """ Take the local config files for Elasticsearch for index, reload them into ES"""
```
Comment (david-fisher) [inline ltr/client/opensearch_client.py]: Nit: OpenSearch, not Elasticsearch

File: notebooks/opensearch/README.md [hunk 7]
```diff
@@ -0,0 +1,6 @@
+This folder contains some Elasticsearch configuration and a Dockerfile to expedite setting up Elasticsearch with LTR.
```
Comment (david-fisher) [inline notebooks/opensearch/README.md]: Nit: OpenSearch not Elasticsearch

**Review by david-fisher — APPROVED** [review]
Very nice. Thank you. A few nits where Elasticsearch wasn't replaced with OpenSearch in the comments.

File: ltr/client/opensearch_client.py (hunk 5 above)
Comment (macohen) [reply ltr/client/opensearch_client.py]: yes! will do.

File: ltr/client/opensearch_client.py (hunk 4 above)
Comment (macohen) [reply ltr/client/opensearch_client.py:49]: will update the root docker-compose.yml to be opensearch-node1. I've tested more with the notebooks/docker-compose.yml

File: docker-compose.yml (hunk 2 above)
Comment (macohen) [reply docker-compose.yml:49]: done

File: docker-compose.yml (hunk 3 above)
Comment (macohen) [reply docker-compose.yml:65]: yes! thanks.

File: ltr/client/opensearch_client.py (hunk 5 above)
Comment (macohen) [reply ltr/client/opensearch_client.py]: updated

File: ltr/client/opensearch_client.py (hunk 6 above)
Comment (macohen) [reply ltr/client/opensearch_client.py]: updated

File: notebooks/opensearch/README.md (hunk 7 above)
Comment (macohen) [reply notebooks/opensearch/README.md]: updated

File: notebooks/opensearch/tmdb/bayesian-optimization.ipynb [hunk 8]
```diff
@@ -0,0 +1,2125 @@
...
+    "    \n",
+    "    average_cumulative_gain = 0.0\n",
+    "    \n",
+    "    for keywords in judgments['keywords'].unique():\n",
+    "        query = json.dumps(es_query).replace(\"{{keywords}}\", keywords)\n",
+    "        query = json.loads(query)\n",
+    "        \n",
+    "        # print(json.dumps(query))\n",
+    "     \n",
+    "        results = client.es.search(index='tmdb', body=query)\n",
```
Comment (mkr) [inline notebooks/opensearch/tmdb/bayesian-optimization.ipynb]:
```suggestion
    "        results = client.opensearch.search(index='tmdb', body=query)\n",
```

**Comment by mkr** [comment]
Thank you @macohen !

**Pull request merged**